discovery/discovery_test.go

```diff
@@ -15,136 +15,568 @@ package discovery
+				{
+					targetGroups: []config.TargetGroup{
+						{
+							Source: "initial1",
+							Targets: []model.LabelSet{
+								{"__instance__": "10.11.122.11:6003"},
+								{"__instance__": "10.11.122.12:6003"},
+								{"__instance__": "10.11.122.13:6003"},
+							},
+						},
+						{
+							Source:  "initial2",
+							Targets: []model.LabelSet{{"__instance__": "10.11.122.14:6003"}},
+						},
+					},
+					interval: 0,
+				},
+				{
+					targetGroups: []config.TargetGroup{
+						{
+							Source:  "initial1",
+							Targets: []model.LabelSet{{"__instance__": "10.11.122.13:6003"}},
+						},
+					},
+					interval: 10,
+				},
+			},
+		},
+		{
+			title: "Three times sync call",
+			updates: []update{
+				{
+					targetGroups: []config.TargetGroup{
+						{
+							Source: "initial1",
+							Targets: []model.LabelSet{
+								{"__instance__": "10.11.122.11:6003"},
+								{"__instance__": "10.11.122.12:6003"},
+								{"__instance__": "10.11.122.13:6003"},
+							},
+						},
+						{
+							Source:  "initial2",
+							Targets: []model.LabelSet{{"__instance__": "10.11.122.14:6003"}},
+						},
+					},
+					interval: 1000,
+				},
+				{
+					targetGroups: []config.TargetGroup{
+						{
+							Source:  "initial1",
+							Targets: []model.LabelSet{{"__instance__": "10.11.122.12:6003"}},
+						},
+						{
+							Source:  "update1",
+							Targets: []model.LabelSet{{"__instance__": "10.11.122.15:6003"}},
+						},
+					},
+					interval: 3000,
+				},
+				{
+					targetGroups: []config.TargetGroup{
+						{
+							Source:  "initial1",
+							Targets: []model.LabelSet{},
+						},
+						{
+							Source: "update1",
+							Targets: []model.LabelSet{
+								{"__instance__": "10.11.122.15:6003"},
+								{"__instance__": "10.11.122.16:6003"},
+							},
+						},
+					},
+					interval: 3000,
+				},
+			},
+		},
+	}
+
+	for i, testCase := range testCases {
+
+		expectedGroups := make(map[string]config.TargetGroup)
+		for _, update := range testCase.updates {
+			for _, targetGroup := range update.targetGroups {
+				expectedGroups[targetGroup.Source] = targetGroup
+			}
+		}
+
+		finalize := make(chan bool)
+
+		targetSet := NewTargetSet(&mockSyncer{
+			sync: func(tgs []*config.TargetGroup) {
+
+				endState := make(map[string]config.TargetGroup)
+				for k, v := range expectedGroups {
+					endState[k] = v
+				}
+
+				if endStateAchieved(tgs, endState) == false {
+					return
+				}
+
+				finalize <- true
+			},
+		})
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		tp := newMockTargetProvider(testCase.updates)
+		targetProviders := map[string]TargetProvider{}
+		targetProviders["testProvider"] = tp
+
+		go targetSet.Run(ctx)
+		targetSet.UpdateProviders(targetProviders)
+
+		select {
+		case <-time.After(20000 * time.Millisecond):
+			t.Errorf("In test case %v[%v]: Test timed out after 20000 millisecond. All targets should be sent within the timeout", i, testCase.title)
```

Comment:
Nit: use `%d` for integers, `%q` for strings. I'd shorten this to `%d. %q: timeout after 20 seconds. ...`